Implemented AST for regular commands

Alex23087 opened the pull request.

Yurand2000 commented:
The module mechanism with the generic annotation type is fitting for the task, but a number of issues are present:
- Names are short and can be confusing, I would suggest having longer names (as an example, make 'rcmd' into 'RegularCommand')
- Both the @@ and @! operators are confusing, prefer functions with names.
- All the constructors for the various expressions and commands belong to the same module (see test_ast_program.ml at the time of writing), while it would be better to distinguish which base type they come from. I know that types are not being mixed with the original approach, but this is always for the sake of readability.

Other comments:
- Is it possible to implement a print function which also prints the opaque annotation type?
- Is it possible to have the ASTRegularCommands module inside the root module instead of it being into Ast_program?

Alex23087 commented:
Addressing these issues in order:

- Agree on the naming, will be changing them exchanging conciseness for clarity;
- Same as previous point;
- That is a good point, working on it.

- In general it is not possible, as the annotation is not necessarily a `Show`able type. A possible design choice is defining a default print function that behaves like `[@opaque]`, and allowing the user to pass a different one. Making this work together with ppx could make the code and the instantiation of the module more clunky;
- Not sure about that, if OCAML defines a module for each file, probably no.
